File: build_matrices.r
suppressPackageStartupMessages(library(lme4))
library(feather)
library(optparse)

option_list <- list(
    make_option("--data"),
    make_option("--formula"),
    make_option("--X"),
    make_option("--Z"),
    make_option("--Lambdatx"),
    make_option("--Lambdati"),
    make_option("--Lambdatp")
)
opt <- parse_args(OptionParser(option_list=option_list))

data <- read_feather(opt$data)
formula <- readLines(opt$formula)

lf <- lFormula(formula, data)

file <- file(opt$X, "wb")
writeBin(c(t(lf$X)), file)
close(file)

file <- file(opt$Z, "wb")
writeBin(c(t(as.matrix(lf$reTrms$Zt))), file)
close(file)

Lambdat <- lf$reTrms$Lambdat
file <- file(opt$Lambdatx, "wb")
writeBin(Lambdat@x, file)
close(file)
file <- file(opt$Lambdati, "wb")
writeBin(Lambdat@i, file)
close(file)
file <- file(opt$Lambdatp, "wb")
writeBin(Lambdat@p, file)
close(file)

File: build_matrices.py
import argparse
import mixed
import pandas as pd

parser = argparse.ArgumentParser()
parser.add_argument('--data')
parser.add_argument('--formula')
parser.add_argument('--X')
parser.add_argument('--Z')
parser.add_argument('--Lambdatx')
parser.add_argument('--Lambdati')
parser.add_argument('--Lambdatp')

args = parser.parse_args()

data = pd.read_feather(args.data)
with open(args.formula, 'r') as file:
    formula = file.read().strip()

X, Z, Lambdat, _, _ = mixed.get_matrices(data, formula)

X.tofile(args.X)
Z.todense().tofile(args.Z)
Lambdat.data.tofile(args.Lambdatx)
Lambdat.indices.tofile(args.Lambdati)
Lambdat.indptr.tofile(args.Lambdatp)
